### PR TITLE
Abstract bonding logic

### PR DESCRIPTION
**Pull request by xlc**
abstract bonding logic so it can be reused by other pallets later

### CODE REVIEW

File: primitives/src/bonding.rs [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+	MinBondThreshold: Get<Balance>,
+{
+	pub fn new() -> Self {
+		Self {
+			unlocking: Default::default(),
+			total: Default::default(),
+			active: Default::default(),
+			_phantom: Default::default(),
+		}
+	}
```
Comment (shaunxw) [inline primitives/src/bonding.rs:79]: I think `new` impl can be moved into `default`, and we always use `default`.

File: modules/nominees-election/src/lib.rs [hunk 2]
```diff
@@ -232,20 +158,18 @@ pub mod module {
...
-				ledger.active += extra;
-				ensure!(
-					ledger.active >= T::MinBondThreshold::get(),
-					Error::<T, I>::BelowMinBondThreshold
-				);
-				ledger.total += extra;
+
+				let old_active = ledger.active();
+
+				let ledger = ledger.bond(extra).map_err(Into::<Error<T, I>>::into)?;
```
Comment (shaunxw) [inline modules/nominees-election/src/lib.rs:168]: IMO we should avoid variable shadowing which is bug prone. `new_ledge` might be better.

File: modules/nominees-election/src/lib.rs [hunk 3]
```diff
@@ -256,30 +180,21 @@ pub mod module {
...
 
-			let mut ledger = Self::ledger(&who);
+			let ledger = Self::ledger(&who);
 
-			let amount = amount.min(ledger.active);
+			let old_active = ledger.active();
+
+			// Note: in case there is no current era it is fine to bond one era more.
+			let era = Self::current_era() + T::BondingDuration::get();
+			let (ledger, amount) = ledger.unbond(amount, era).map_err(Into::<Error<T, I>>::into)?;
```
Comment (shaunxw) [inline modules/nominees-election/src/lib.rs:189]: Same here.

File: modules/nominees-election/src/lib.rs [hunk 4]
```diff
@@ -288,35 +203,35 @@ pub mod module {
...
 			let ledger = Self::ledger(&who);
-			ensure!(!ledger.unlocking.is_empty(), Error::<T, I>::NoUnlockChunk);
-			let old_active = ledger.active;
-			let old_ledger_unlocking = ledger.unlocking.len();
+
+			let old_active = ledger.active();
+			let old_ledger_unlocking = ledger.unlocking_len();
 			let old_nominations = Self::nominations(&who);
-			let ledger = ledger.rebond(amount);
+			let (ledger, amount) = ledger.rebond(amount).map_err(Into::<Error<T, I>>::into)?;
```
Comment (shaunxw) [inline modules/nominees-election/src/lib.rs:210]: and here

File: modules/nominees-election/src/lib.rs [hunk 5]
```diff
@@ -288,35 +203,35 @@ pub mod module {
...
 		#[pallet::weight(T::WeightInfo::withdraw_unbonded(T::MaxUnlockingChunks::get()))]
 		#[transactional]
 		pub fn withdraw_unbonded(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			let mut ledger = Self::ledger(&who);
-			let old_ledger_unlocking = ledger.unlocking.len();
-			ledger.consolidate_unlocked(Self::current_era());
+			let ledger = Self::ledger(&who);
+			let old_ledger_unlocking = ledger.unlocking_len();
+			let ledger = ledger.consolidate_unlocked(Self::current_era());
```
Comment (shaunxw) [inline modules/nominees-election/src/lib.rs:225]: and here

**Review by shaunxw — COMMENTED** [review]
LGTM, some nitpicks

File: modules/nominees-election/src/lib.rs (hunk 2 above)
Comment (xlc) [reply modules/nominees-election/src/lib.rs:168]: It is actually better to do shadowing than `ledger1`, `ledger2`, `ledger3` which you will eventually lost track.
It basically implements mutation without mutation.

**Pull request merged**